File: main/lru-cache-lc/lru-cache-lc-2-scratch.cpp
#include <cassert>
#include <iostream>
#include <iterator>
#include <list>
#include <stdexcept>
#include <unordered_map>

using namespace std;

class LRUCache {
public:
    static constexpr auto not_found = -1;

    explicit LRUCache(int capacity);

    int get(int key) noexcept;

    void put(int key, int value);

private:
    struct KeyValuePair {
        int key, value;

        KeyValuePair(const int key, const int value)
            : key{key}, value{value} { }
    };

    void bump(const list<KeyValuePair>::const_iterator pos) // splices to front
    {
        chain_.splice(cbegin(chain_), chain_, pos);
    }

    unordered_map<int, list<KeyValuePair>::iterator> lookup_;
    list<KeyValuePair> chain_;
    const list<KeyValuePair>::size_type capacity_;
};

LRUCache::LRUCache(const int capacity)
    : capacity_{static_cast<list<KeyValuePair>::size_type>(capacity)}
{
    if (capacity <= 0)
        throw invalid_argument{"capacity must be strictly positive"};
}

int LRUCache::get(const int key) noexcept
{
    const auto it = lookup_.find(key);
    if (it == end(lookup_)) return not_found;

    const auto pos = it->second;
    bump(pos);
    return pos->value;
}

void LRUCache::put(const int key, const int value)
{
    const auto it = lookup_.find(key);

    if (it != end(lookup_)) {
        const auto pos = it->second;
        pos->value = value;
        bump(pos);
    }
    else if (chain_.size() == capacity_) {
        const auto pos = prev(end(chain_)); // reuse back (least recent) node
        lookup_.erase(pos->key);

        pos->key = key;
        pos->value = value;
        bump(pos);
        lookup_.emplace(key, pos);
    } else {
        assert(chain_.size() < capacity_);

        chain_.emplace_front(key, value);
        lookup_.emplace(key, begin(chain_));
    }
}

int main()
{
    LRUCache cache{2};

    cache.put(1, 1);
    cache.put(2, 2);
    cout << cache.get(1) << '\n';
    cache.put(3, 3);
    cout << cache.get(2) << '\n';
    cache.put(4, 4);
    cout << cache.get(1) << '\n';
    cout << cache.get(3) << '\n';
    cout << cache.get(4) << '\n';
}

File: main/lru-cache-lc/lru-cache-lc-2-scratch.java
import java.util.HashMap;
import java.util.Map;

/** A node in a circular doubly-linked list. */
final class Node {
    /** Creates a node suitable for use as a sentinel. */
    Node() { }

    /** Creates a node suitable for use as a data node. */
    Node(final int key, final int value) {
        _key = key;
        _value = value;
    }

    /** Gets the node that comes before this one. */
    Node prev() { return _prev; }

    /** Gets the node that comes after this one. */
    Node next() { return _next; }

    /** Gets the key held by this node. */
    int key() { return _key; }

    /** Changes the key held by this node. */
    void setKey(final int key) { _key = key; }

    /** Gets the value held by this node. */
    int value() { return _value; }

    /** Changes the value held by this node. */
    void setValue(final int value) { _value = value; }

    /** Disconnect this node from its adjacent nodes, if any. */
    void detach() {
        _prev._next = _next;
        _next._prev = _prev;

        _prev = _next = this;
    }

    /** Detach the other node and (re)attach it just after this one. */
    void spliceNext(final Node other) {
        other.detach();

        other._prev = this;
        other._next = _next;

        _next._prev = other;
        _next = other;
    }

    private Node _prev = this, _next = this;

    private int _key, _value;
}

class LRUCache {
    static final int NOT_FOUND = -1;

    public LRUCache(final int capacity) {
        if (capacity <= 0) {
            throw new IllegalArgumentException(
                    "capacity must be strictly positive");
        }

        _capacity = capacity;
    }

    public int get(final int key) {
        final Node node = _map.get(key);
        if (node == null) return NOT_FOUND;

        _list.spliceNext(node);
        return node.value();
    }

    public void put(final int key, final int value) {
        Node node = _map.get(key);

        if (node != null) {
            node.setValue(value);
            _list.spliceNext(node);
        }
        else if (_size == _capacity) {
            node = _list.prev(); // reuse the least recently used node
            _map.remove(node.key());

            node.setKey(key);
            node.setValue(value);
            _list.spliceNext(node);
            _map.put(key, node);
        } else {
            assert _size < _capacity;

            node = new Node(key, value);
            _list.spliceNext(node);
            _map.put(key, node);
            ++_size;
        }

    }

    private final Node _list = new Node();

    private final Map<Integer, Node> _map = new HashMap<>();

    private final int _capacity;

    private int _size = 0;
}

final class UnitTest {
    public static void main(final String[] args) {
        final LRUCache cache = new LRUCache(2);

        cache.put(1, 1);
        cache.put(2, 2);
        p(cache.get(1));
        cache.put(3, 3);
        p(cache.get(2));
        cache.put(4, 4);
        p(cache.get(1));
        p(cache.get(3));
        p(cache.get(4));
    }

    private static void p(final Object obj) {
        System.out.println(obj);
    }

    private UnitTest() { throw new AssertionError(); }
}
